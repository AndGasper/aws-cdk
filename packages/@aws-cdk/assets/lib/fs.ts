import crypto = require('crypto');
import fs = require('fs');
import minimatch = require('minimatch');
import path = require('path');

const BUFFER_SIZE = 8 * 1024;

export interface CopyOptions {
  /**
   * @default External only follows symlinks that are external to the source directory
   */
  follow?: FollowMode;

  /**
   * glob patterns to exclude from the copy.
   */
  exclude?: string[];
}

export enum FollowMode {
  /**
   * Never follow symlinks.
   */
  Never = 'never',

  /**
   * Materialize all symlinks, whether they are internal or external to the source directory.
   */
  Always = 'always',

  /**
   * Only follows symlinks that are external to the source directory.
   */
  External = 'external',

  // ----------------- TODO::::::::::::::::::::::::::::::::::::::::::::
  /**
   * Forbids source from having any symlinks pointing outside of the source
   * tree.
   *
   * This is the safest mode of operation as it ensures that copy operations
   * won't materialize files from the user's file system. Internal symlinks are
   * not followed.
   *
   * If the copy operation runs into an external symlink, it will fail.
   */
  BlockExternal = 'internal-only',
}

export function copyDirectory(srcDir: string, destDir: string, options: CopyOptions = { }, rootDir?: string) {
  const follow = options.follow !== undefined ? options.follow : FollowMode.External;
  const exclude = options.exclude || [];

  rootDir = rootDir || srcDir;

  if (!fs.statSync(srcDir).isDirectory()) {
    throw new Error(`${srcDir} is not a directory`);
  }

  const files = fs.readdirSync(srcDir);
  for (const file of files) {
    const sourceFilePath = path.join(srcDir, file);

    if (shouldExclude(path.relative(rootDir, sourceFilePath))) {
      continue;
    }

    const destFilePath = path.join(destDir, file);

    let stat: fs.Stats | undefined = follow === FollowMode.Always
      ? fs.statSync(sourceFilePath)
      : fs.lstatSync(sourceFilePath);

    if (stat && stat.isSymbolicLink()) {
      const target = fs.readlinkSync(sourceFilePath);

      // determine if this is an external link (i.e. the target's absolute path
      // is outside of the root directory).
      const targetPath = path.normalize(path.resolve(srcDir, target));
      const rootPath = path.normalize(rootDir);
      const external = !targetPath.startsWith(rootPath);

      if (follow === FollowMode.External && external) {
        stat = fs.statSync(sourceFilePath);
      } else {
        fs.symlinkSync(target, destFilePath);
        stat = undefined;
      }
    }

    if (stat && stat.isDirectory()) {
      fs.mkdirSync(destFilePath);
      copyDirectory(sourceFilePath, destFilePath, options, rootDir);
      stat = undefined;
    }

    if (stat && stat.isFile()) {
      fs.copyFileSync(sourceFilePath, destFilePath);
      stat = undefined;
    }
  }

  function shouldExclude(filePath: string): boolean {
    let excludeOutput = false;

    for (const pattern of exclude) {
      const negate = pattern.startsWith('!');
      const match = minimatch(filePath, pattern, { matchBase: true, flipNegate: true });

      if (!negate && match) {
        excludeOutput = true;
      }

      if (negate && match) {
        excludeOutput = false;
      }
    }

    return excludeOutput;
  }
}

export function fingerprint(fileOrDirectory: string) {
  const hash = crypto.createHash('md5');
  addToHash(fileOrDirectory);
  return hash.digest('hex');

  function addToHash(pathToAdd: string) {
    hash.update('==\n');
    const relativePath = path.relative(fileOrDirectory, pathToAdd);
    hash.update(relativePath + '\n');
    hash.update('~~~~~~~~~~~~~~~~~~\n');
    const stat = fs.statSync(pathToAdd);

    if (stat.isSymbolicLink()) {
      const target = fs.readlinkSync(pathToAdd);
      hash.update(target);
    } else if (stat.isDirectory()) {
      for (const file of fs.readdirSync(pathToAdd)) {
        addToHash(path.join(pathToAdd, file));
      }
    } else {
      const file = fs.openSync(pathToAdd, 'r');
      const buffer = Buffer.alloc(BUFFER_SIZE);

      try {
        let bytesRead;
        do {
          bytesRead = fs.readSync(file, buffer, 0, BUFFER_SIZE, null);
          hash.update(buffer.slice(0, bytesRead));
        } while (bytesRead === BUFFER_SIZE);
      } finally {
        fs.closeSync(file);
      }
    }
  }
}