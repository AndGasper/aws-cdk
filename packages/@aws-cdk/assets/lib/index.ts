export * from './asset';
export * from './follow-mode';