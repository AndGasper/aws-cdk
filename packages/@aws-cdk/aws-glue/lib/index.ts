// AWS::Glue CloudFormation Resources:
export * from './glue.generated';

export * from './database';
export * from './schema';
export * from './storage-type';
export * from './table';