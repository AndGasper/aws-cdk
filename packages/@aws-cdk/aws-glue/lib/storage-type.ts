export enum InputFormat {
  TextInputFormat = 'org.apache.hadoop.mapred.TextInputFormat'
}

export enum OutputFormat {
  HiveIgnoreKeyTextOutputFormat = 'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat'
}

export enum SerializationLibrary {
  OpenXJson = 'org.openx.data.jsonserde.JsonSerDe'
}

export interface StorageType {
  inputFormat: string;
  outputFormat: string;
  serializationLibrary: string;
}

export namespace StorageType {
  export const Json: StorageType = {
    inputFormat: InputFormat.TextInputFormat,
    outputFormat: OutputFormat.HiveIgnoreKeyTextOutputFormat,
    serializationLibrary: SerializationLibrary.OpenXJson
  };
}