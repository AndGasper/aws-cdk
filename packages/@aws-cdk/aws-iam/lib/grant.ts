import cdk = require('@aws-cdk/cdk');
import { PolicyStatement } from "./policy-document";
import { IPrincipal } from "./principals";

/**
 * Options for a grant operation
 */
export interface GrantOptions {
  /**
   * The principal to grant to
   *
   * @default No work is done
   */
  principal: IPrincipal | undefined;

  /**
   * The actions to grant
   */
  actions: string[];

  /**
   * The resource ARNs to grant to
   */
  resourceArns: string[];

  /**
   * Adder to the resource policy
   *
   * Either 'scope' or 'resource' must be supplied.
   *
   * An error will be thrown if the policy could not be added to the principal,
   * no resource is supplied given and `skipResourcePolicy` is false.
   */
  resource?: IResourceWithPolicy;

  /**
   * When referring to the resource in a resource policy, use this as ARN.
   *
   * (Depending on the resource type, this needs to be '*' in a resource policy).
   *
   * @default Same as regular resource ARNs
   */
  resourceSelfArns?: string[];

  /**
   * Construct to report warnings on in case grant could not be registered
   *
   * @default resource
   */
  scope?: cdk.IConstruct;
}

/**
 * Options for a tryGrant operation
 */
export interface TryGrantOnIdentityOptions {
  /**
   * The principal to grant to
   *
   * @default No work is done
   */
  principal: IPrincipal | undefined;

  /**
   * The actions to grant
   */
  actions: string[];

  /**
   * The resource ARNs to grant to
   */
  resourceArns: string[];

  /**
   * Construct to report warnings on in case grant could not be registered
   */
  scope: cdk.IConstruct;
}

/**
 * Options for a grant operation to both identity and resource
 */
export interface GrantOnIdentityAndResourceOptions {
  /**
   * The principal to grant to
   *
   * @default No work is done
   */
  principal: IPrincipal | undefined;

  /**
   * The actions to grant
   */
  actions: string[];

  /**
   * The resource ARNs to grant to
   */
  resourceArns: string[];

  /**
   * Adder to the resource policy
   *
   * Either 'scope' or 'resource' must be supplied.
   *
   * An error will be thrown if the policy could not be added to the principal,
   * no resource is supplied given and `skipResourcePolicy` is false.
   */
  resource: IResourceWithPolicy;

  /**
   * When referring to the resource in a resource policy, use this as ARN.
   *
   * (Depending on the resource type, this needs to be '*' in a resource policy).
   *
   * @default Same as regular resource ARNs
   */
  resourceSelfArns?: string[];
}

export class Permissions {
  /**
   * Grant the given permissions to the principal
   *
   * The permissions will be added to the principal policy primarily, falling
   * back to the resource policy if necessary. The permissions must be granted
   * somewhere.
   *
   * - Trying to grant permissions to a principal that does not admit adding to
   *   the principal policy while not providing a resource with a resource policy
   *   is an error.
   * - Trying to grant permissions to an absent principal (possible in the
   *   case of imported resources) leads to a warning being added to the
   *   resource construct.
   */
  public static grant(options: GrantOptions): GrantResult {
    const scope = options.scope || options.resource;
    if (!scope) {
      throw new Error(`Either 'scope' or 'resource' must be supplied.`);
    }

    const result = Permissions.tryGrantOnPrincipal({
      actions: options.actions,
      principal: options.principal,
      resourceArns: options.resourceArns,
      scope
    });

    if (result.addedToPrincipal || result.principalMissing) { return result; }

    if (!options.resource) {
      // tslint:disable-next-line:max-line-length
      throw new Error(`Neither principal (${options.principal}) nor resource (${scope}) allow adding to policy. Grant to a Role instead.`);
    }

    const statement = new PolicyStatement()
      .addActions(...options.actions)
      .addResources(...(options.resourceSelfArns || options.resourceArns))
      .addPrincipal(options.principal!);

    options.resource.addToResourcePolicy(statement);

    return grantResult({ addedToResource: true, statement });
  }

  /**
   * Try to grant the given permissions to the given principal
   *
   * Absence of a principal leads to a warning, but failing to add
   * the permissions to a present principal is not an error.
   */
  public static tryGrantOnPrincipal(options: TryGrantOnIdentityOptions): GrantResult {
    if (!options.principal) {
      // tslint:disable-next-line:max-line-length
      options.scope.node.addWarning(`Could not add grant for '${options.actions}' on '${options.resourceArns}' because the principal was not available. Add the permissions by hand.`);
      return grantResult({ principalMissing: true });
    }

    const statement = new PolicyStatement()
      .addActions(...options.actions)
      .addResources(...options.resourceArns);

    const addedToPrincipal = options.principal.addToPolicy(statement);

    return grantResult({ addedToPrincipal, statement: addedToPrincipal ? statement : undefined });
  }

  /**
   * Add a grant both on the principal and on the resource
   *
   * As long as any principal is given, granting on the pricipal may fail (in
   * case of a non-identity principal), but granting on the resource will
   * always be done.
   *
   * Statement will be the resource statement.
   */
  public static grantOnPrincipalAndResource(options: GrantOnIdentityAndResourceOptions): GrantResult {
    const result = Permissions.tryGrantOnPrincipal({
      actions: options.actions,
      principal: options.principal,
      resourceArns: options.resourceArns,
      scope: options.resource
    });

    if (result.principalMissing) { return result; }

    const statement = new PolicyStatement()
      .addActions(...options.actions)
      .addResources(...(options.resourceSelfArns || options.resourceArns))
      .addPrincipal(options.principal!);

    options.resource.addToResourcePolicy(statement);

    return grantResult({ addedToPrincipal: result.addedToPrincipal, addedToResource: true, statement });
  }
}

/**
 * The result of the grant() operation
 *
 * This class is not instantiable by consumers on purpose, so that they will be
 * required to call the Permissions.grant() functions.
 */
export class GrantResult {
  /**
   * There was no principal to add the permissions to
   */
  public readonly principalMissing: boolean;

  /**
   * The grant was added to the principal's policy
   */
  public readonly addedToPrincipal: boolean;

  /**
   * The grant was added to the resource policy
   */
  public readonly addedToResource: boolean;

  /**
   * The policy statement that was created for this permission
   *
   * Can be accessed to (e.g.) add additional conditions to the statement.
   *
   * Only set if either addedToPrincipal or addedToResource are true.
   */
  public readonly statement?: PolicyStatement;

  /**
   * Private member to make it impossible to construct object literals that structurally match this type
   */
  private readonly _isGrantResult = true;

  private constructor(props: GrantResultProps) {
    this.principalMissing = !!props.principalMissing;
    this.addedToPrincipal = !!props.addedToPrincipal;
    this.addedToResource = !!props.addedToResource;
    this.statement = props.statement;
    Array.isArray(this._isGrantResult);
  }
}

/**
 * Instantiate a grantResult (which is normally not instantiable)
 */
function grantResult(props: GrantResultProps): GrantResult {
  return Reflect.construct(GrantResult, [props]);
}

interface GrantResultProps {
  principalMissing?: boolean;
  addedToPrincipal?: boolean;
  addedToResource?: boolean;
  statement?: PolicyStatement;
}

/**
 * A resource with a resource policy that can be added to
 */
export interface IResourceWithPolicy extends cdk.IConstruct {
  /**
   * Add a statement to the resource's resource policy
   */
  addToResourcePolicy(statement: PolicyStatement): void;
}