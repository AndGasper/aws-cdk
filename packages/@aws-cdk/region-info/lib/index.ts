export * from './region-info-api';
export * from './region-info';
