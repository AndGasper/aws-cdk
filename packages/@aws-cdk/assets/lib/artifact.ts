import cdk = require('@aws-cdk/cdk');
import cxapi = require('@aws-cdk/cx-api');
import fs = require('fs');
import path = require('path');
import { AssetPackaging, GenericAssetProps } from './asset';
import { copyDirectory } from './fs-copy';
import { fingerprint } from './fs-fingerprint';

export interface AssetArtifactProps extends GenericAssetProps {
  fingerprint: string;
}

/**
 * This is an internal construct (only available within this module) which represents a unique
 * asset within the app.
 *
 * To get an instance, use `AssetArtifact.forAsset`. It will calculate the fingerprint
 * of the asset based on content hash and look up if there is already an artifact with this
 * fingerprint in the app.
 */
export class AssetArtifact extends cdk.Construct implements cdk.ISynthesizable {

  /**
   * @returns gets or creates a singletone asset artifact for a specific asset
   */
  public static forAsset(scope: cdk.Construct, props: GenericAssetProps): AssetArtifact {
    const extra = {
      packaging: props.packaging,
      extra: props.extra,
    };

    const sourcePath = path.resolve(props.path);
    validateAssetOnDisk(sourcePath, props.packaging);

    // calculate content hash, which is what we use as the app-level ID of this asset
    const fp = fingerprint(props.path, {
      exclude: props.exclude,
      follow: props.follow,
      extra: JSON.stringify(extra)
    });

    const root = scope.node.root;

    const id = `asset_${fp}`;
    const artifact = root.node.tryFindChild(id) as AssetArtifact || new AssetArtifact(root, id, {
      fingerprint: fp,
      path: sourcePath,
      ...props
    });

    return artifact;
  }

  /**
   * The unique fingerprint of this asset, based MD5 hash of the content and the
   * directory structure.
   */
  public readonly fingerprint: string;

  private readonly artifactId: string;
  private readonly environments = new Set<string>();

  private constructor(scope: cdk.Construct, id: string, private readonly props: AssetArtifactProps) {
    super(scope, id);

    this.fingerprint = props.fingerprint;
    this.artifactId = id; // based on the fingerprint
  }

  /**
   * Creates (or reuses) the set of bucket/key parameters that will be used to wire assets
   * into a specific stack. Called by the `Asset` class to "subscribe" the asset to the stack.
   *
   * @param stack
   */
  public wireToStack(stack: cdk.Stack): AssetArtifactParameters {
    const bucketParamId = `asset-${this.fingerprint}-S3Bucket`;
    const keyParamId = `asset-${this.fingerprint}-S3Key`;
    const hashParamId = `asset-${this.fingerprint}-S3ContentHash`;

    // if we already have a parameters wired for this asset, just return them (assuming they all exist)
    if (stack.node.tryFindChild(keyParamId)) {
      return {
        bucket: stack.node.findChild(bucketParamId) as cdk.Parameter,
        key: stack.node.findChild(keyParamId) as cdk.Parameter,
        sha256: stack.node.findChild(hashParamId) as cdk.Parameter
      };
    }

    // okay, "register" this asset with the stack. this includes adding a bunch of parameters
    // and a metadata entry that tells the toolkit how to build the asset.

    const bucketParam = new cdk.Parameter(stack, bucketParamId, {
      type: 'String',
      description: `S3 bucket for asset ${this.fingerprint} from ${this.props.path}`,
    });

    const keyParam = new cdk.Parameter(stack, keyParamId, {
      type: 'String',
      description: `S3 key for asset ${this.fingerprint} from ${this.props.path}`
    });

    const contentHashParam = new cdk.Parameter(stack, hashParamId, {
      type: 'String',
      description: 'SHA-256 hash of the asset as it was uploaded to S3'
    });

    stack.setParameterValue(bucketParam, `\${${this.artifactId}}.s3bucket`);
    stack.setParameterValue(keyParam, `\${${this.artifactId}}.s3key`);
    stack.setParameterValue(contentHashParam, `\${${this.artifactId}}.sha256`);

    this.environments.add(stack.environment);

    return {
      bucket: bucketParam,
      key: keyParam,
      sha256: contentHashParam
    };
  }

  public synthesize(session: cdk.ISynthesisSession) {

    switch (this.props.packaging) {
      case AssetPackaging.File:
        this.synthesizeFile(session);
        return;

      case AssetPackaging.ZipDirectory:
        this.synthesizeZipDirectory(session);
        return;
    }

  }

  private synthesizeFile(session: cdk.ISynthesisSession) {
    const ext = path.extname(this.props.path);
    const fileName = `${this.fingerprint}${ext}`;

    const stagingFile = path.join(this.fingerprint, path.basename(this.props.path));

    fs.mkdirSync(path.join(session.staging.path, this.fingerprint));
    fs.copyFileSync(this.props.path, path.join(session.staging.path, stagingFile));

    session.addBuildStep(this.artifactId, {
      type: 'CopyFileTask',
      parameters: {
        src: `staging/${stagingFile}`,
        dest: `assembly/${fileName}` // relative to outdir
      }
    });

    for (const env of this.environments) {
      session.addArtifact(this.artifactId, {
        type: cxapi.ArtifactType.AwsS3Object,
        environment: env,
        properties: {
          file: fileName // relative to assembly
        }
      });
    }
  }

  private synthesizeZipDirectory(session: cdk.ISynthesisSession) {
    const fileName = `${this.fingerprint}.zip`;

    // copy source contents to a staging directory
    const stagingDir = this.fingerprint;
    fs.mkdirSync(path.join(session.staging.path, stagingDir));
    copyDirectory(this.props.path, path.join(session.staging.path, stagingDir));

    session.addBuildStep(this.artifactId, {
      type: 'ZipDirectoryTask',
      parameters: {
        src: `staging/${stagingDir}`,
        dest: `assembly/${fileName}`, // relative to outdir
      }
    });

    for (const env of this.environments) {
      session.addArtifact(this.artifactId, {
        type: cxapi.ArtifactType.AwsS3Object,
        environment: env,
        properties: {
          file: fileName // relative to assembly/
        }
      });
    }
  }
}

export interface AssetArtifactParameters {
  bucket: cdk.Parameter;
  key: cdk.Parameter;
  sha256: cdk.Parameter;
}

function validateAssetOnDisk(assetPath: string, packaging: AssetPackaging) {
  if (!fs.existsSync(assetPath)) {
    throw new Error(`Cannot find asset at ${assetPath}`);
  }

  switch (packaging) {
    case AssetPackaging.ZipDirectory:
      if (!fs.statSync(assetPath).isDirectory()) {
        throw new Error(`${assetPath} is expected to be a directory when asset packaging is 'zip'`);
      }
      break;

    case AssetPackaging.File:
      if (!fs.statSync(assetPath).isFile()) {
        throw new Error(`${assetPath} is expected to be a regular file when asset packaging is 'file'`);
      }
      break;

    default:
      throw new Error(`Unsupported asset packaging format: ${packaging}`);
  }
}
