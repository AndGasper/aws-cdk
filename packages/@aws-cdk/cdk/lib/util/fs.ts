import crypto = require('crypto');
import fs = require('fs');
import path = require('path');

const BUFFER_SIZE = 8 * 1024;

export interface CopyOptions {
  /**
   * @default External only follows symlinks that are external to the source directory
   */
  follow?: FollowMode;
}

export enum FollowMode {
  Never = 'never',
  Always = 'always',

  /**
   * Only follows symlinks that are external to the source directory.
   */
  External = 'external'
}

export function copyFile(src: string, dest: string, options?: CopyOptions) {
  fs.copyFileSync(src, dest);
}

export function copyDirectory(srcDir: string, destDir: string, options: CopyOptions = { }, rootDir?: string) {
  const follow = options.follow !== undefined ? options.follow : FollowMode.External;

  rootDir = rootDir || srcDir;

  if (!fs.statSync(srcDir).isDirectory()) {
    throw new Error(`${srcDir} is not a directory`);
  }

  const files = fs.readdirSync(srcDir);
  for (const file of files) {
    const sourceFilePath = path.join(srcDir, file);
    const destFilePath = path.join(destDir, file);

    let stat: fs.Stats | undefined = follow === FollowMode.Always
      ? fs.statSync(sourceFilePath)
      : fs.lstatSync(sourceFilePath);

    if (stat && stat.isSymbolicLink()) {
      const target = fs.readlinkSync(sourceFilePath);

      // determine if this is an external link (i.e. the target's absolute path
      // is outside of the root directory).
      const targetPath = path.normalize(path.resolve(srcDir, target));
      const rootPath = path.normalize(rootDir);
      const external = !targetPath.startsWith(rootPath);

      if (follow === FollowMode.External && external) {
        stat = fs.statSync(sourceFilePath);
      } else {
        fs.symlinkSync(target, destFilePath);
        stat = undefined;
      }
    }

    if (stat && stat.isDirectory()) {
      fs.mkdirSync(destFilePath);
      copyDirectory(sourceFilePath, destFilePath, options, rootDir);
      stat = undefined;
    }

    if (stat && stat.isFile()) {
      fs.copyFileSync(sourceFilePath, destFilePath);
      stat = undefined;
    }
  }
}

export function fingerprint(fileOrDirectory: string) {
  const hash = crypto.createHash('md5');
  addToHash(fileOrDirectory);
  return hash.digest('hex');

  function addToHash(pathToAdd: string) {
    hash.update('==\n');
    const relativePath = path.relative(fileOrDirectory, pathToAdd);
    hash.update(relativePath + '\n');
    hash.update('~~~~~~~~~~~~~~~~~~\n');
    const stat = fs.statSync(pathToAdd);

    if (stat.isSymbolicLink()) {
      const target = fs.readlinkSync(pathToAdd);
      hash.update(target);
    } else if (stat.isDirectory()) {
      for (const file of fs.readdirSync(pathToAdd)) {
        addToHash(path.join(pathToAdd, file));
      }
    } else {
      const file = fs.openSync(pathToAdd, 'r');
      const buffer = Buffer.alloc(BUFFER_SIZE);

      try {
        let bytesRead;
        do {
          bytesRead = fs.readSync(file, buffer, 0, BUFFER_SIZE, null);
          hash.update(buffer.slice(0, bytesRead));
        } while (bytesRead === BUFFER_SIZE);
      } finally {
        fs.closeSync(file);
      }
    }
  }
}