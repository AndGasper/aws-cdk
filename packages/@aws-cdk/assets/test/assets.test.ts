import cdk = require('@aws-cdk/cdk');
import cxapi = require('@aws-cdk/cx-api');
import fs = require('fs');
import path = require('path');
import { FileAsset, ZipDirectoryAsset } from '../lib';

const SAMPLE_ASSET_DIRECTORY = path.join(__dirname, 'fixtures', 'test1');

test('file assets', () => {
  // GIVEN
  const app = new cdk.App({
    context: {
      [cxapi.DISABLE_STACK_TRACE]: 'true',
      [cxapi.DISABLE_VERSION_REPORTING]: 'true'
    }
  });
  const stack = new cdk.Stack(app, 'stack1');

  // WHEN
  new FileAsset(stack, 'my-first-modern-file-asset', {
    path: path.join(SAMPLE_ASSET_DIRECTORY, 'file1.txt')
  });

  // THEN
  const session = app.run();
  expect(snapshotDirectory(session.path)).toMatchSnapshot();
});

test('zip directory assets', () => {
  // GIVEN
  const app = new cdk.App({ context: { [cxapi.DISABLE_STACK_TRACE]: 'true' } });
  const stack = new cdk.Stack(app, 'stack1');

  // WHEN
  new ZipDirectoryAsset(stack, 'my-first-zip-dir-asset', {
    path: SAMPLE_ASSET_DIRECTORY
  });

  // THEN
  expect(snapshotDirectory(app.run().path)).toMatchSnapshot();
});

test('two assets with the same source will result in a single artifact', () => {
  // GIVEN
  const app = new cdk.App({ context: { [cxapi.DISABLE_STACK_TRACE]: 'true' } });
  const stack = new cdk.Stack(app, 'stack1');

  // WHEN
  new ZipDirectoryAsset(stack, 'asset1', { path: SAMPLE_ASSET_DIRECTORY });
  new ZipDirectoryAsset(stack, 'asset2', { path: SAMPLE_ASSET_DIRECTORY });

  // THEN
  expect(snapshotDirectory(app.run().path)).toMatchSnapshot();
});

function snapshotDirectory(dir: string) {
  const entries: { [name: string]: any } = { };
  for (const file of fs.readdirSync(dir)) {
    const fp = path.join(dir, file);
    if (fs.statSync(fp).isDirectory()) {
      entries[file] = snapshotDirectory(fp);
    } else if (path.extname(file) === '.json') {
      entries[file] = JSON.parse(fs.readFileSync(fp).toString());
    } else {
      entries[file] = fs.readFileSync(fp).toString().trim();
    }
  }
  return entries;
}