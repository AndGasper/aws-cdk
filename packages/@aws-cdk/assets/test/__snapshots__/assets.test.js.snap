// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file assets 1`] = `
Object {
  "assembly": Object {
    "manifest.json": Object {
      "artifacts": Object {
        "asset_4be9b497d3aade053d066e81efc93fbd": Object {
          "environment": "aws://unknown-account/unknown-region",
          "properties": Object {
            "file": "4be9b497d3aade053d066e81efc93fbd.txt",
          },
          "type": "aws:s3:object",
        },
        "stack1": Object {
          "environment": "aws://unknown-account/unknown-region",
          "properties": Object {
            "parameters": Object {
              "asset4be9b497d3aade053d066e81efc93fbdS3Bucket": "\${asset_4be9b497d3aade053d066e81efc93fbd}.s3bucket",
              "asset4be9b497d3aade053d066e81efc93fbdS3ContentHash": "\${asset_4be9b497d3aade053d066e81efc93fbd}.sha256",
              "asset4be9b497d3aade053d066e81efc93fbdS3Key": "\${asset_4be9b497d3aade053d066e81efc93fbd}.s3key",
            },
            "templateFile": "stack1.template.json",
          },
          "type": "aws:cloudformation:stack",
        },
      },
      "version": "0.19.0",
    },
    "stack1.template.json": Object {
      "Parameters": Object {
        "asset4be9b497d3aade053d066e81efc93fbdS3Bucket": Object {
          "Description": "S3 bucket for asset 4be9b497d3aade053d066e81efc93fbd from /Users/benisrae/code/cdk/aws-cdk-3/packages/@aws-cdk/assets/test/fixtures/test1/file1.txt",
          "Type": "String",
        },
        "asset4be9b497d3aade053d066e81efc93fbdS3ContentHash": Object {
          "Description": "SHA-256 hash of the asset as it was uploaded to S3",
          "Type": "String",
        },
        "asset4be9b497d3aade053d066e81efc93fbdS3Key": Object {
          "Description": "S3 key for asset 4be9b497d3aade053d066e81efc93fbd from /Users/benisrae/code/cdk/aws-cdk-3/packages/@aws-cdk/assets/test/fixtures/test1/file1.txt",
          "Type": "String",
        },
      },
    },
  },
  "build.json": Object {
    "steps": Object {
      "asset_4be9b497d3aade053d066e81efc93fbd": Object {
        "parameters": Object {
          "dest": "assembly/4be9b497d3aade053d066e81efc93fbd.txt",
          "src": "staging/4be9b497d3aade053d066e81efc93fbd/file1.txt",
        },
        "type": "CopyFileTask",
      },
    },
  },
  "staging": Object {
    "4be9b497d3aade053d066e81efc93fbd": Object {
      "file1.txt": "file1",
    },
  },
}
`;

exports[`two assets with the same source will result in a single artifact 1`] = `
Object {
  "assembly": Object {
    "manifest.json": Object {
      "artifacts": Object {
        "asset_6a2bd71840ef97020185c6357b101a70": Object {
          "environment": "aws://unknown-account/unknown-region",
          "properties": Object {
            "file": "6a2bd71840ef97020185c6357b101a70.zip",
          },
          "type": "aws:s3:object",
        },
        "stack1": Object {
          "environment": "aws://unknown-account/unknown-region",
          "properties": Object {
            "parameters": Object {
              "asset6a2bd71840ef97020185c6357b101a70S3Bucket": "\${asset_6a2bd71840ef97020185c6357b101a70}.s3bucket",
              "asset6a2bd71840ef97020185c6357b101a70S3ContentHash": "\${asset_6a2bd71840ef97020185c6357b101a70}.sha256",
              "asset6a2bd71840ef97020185c6357b101a70S3Key": "\${asset_6a2bd71840ef97020185c6357b101a70}.s3key",
            },
            "templateFile": "stack1.template.json",
          },
          "type": "aws:cloudformation:stack",
        },
      },
      "runtime": Object {
        "libraries": Object {
          "jsii-runtime": "node.js/v10.13.0",
        },
      },
      "version": "0.19.0",
    },
    "stack1.template.json": Object {
      "Parameters": Object {
        "asset6a2bd71840ef97020185c6357b101a70S3Bucket": Object {
          "Description": "S3 bucket for asset 6a2bd71840ef97020185c6357b101a70 from /Users/benisrae/code/cdk/aws-cdk-3/packages/@aws-cdk/assets/test/fixtures/test1",
          "Type": "String",
        },
        "asset6a2bd71840ef97020185c6357b101a70S3ContentHash": Object {
          "Description": "SHA-256 hash of the asset as it was uploaded to S3",
          "Type": "String",
        },
        "asset6a2bd71840ef97020185c6357b101a70S3Key": Object {
          "Description": "S3 key for asset 6a2bd71840ef97020185c6357b101a70 from /Users/benisrae/code/cdk/aws-cdk-3/packages/@aws-cdk/assets/test/fixtures/test1",
          "Type": "String",
        },
      },
    },
  },
  "build.json": Object {
    "steps": Object {
      "asset_6a2bd71840ef97020185c6357b101a70": Object {
        "parameters": Object {
          "dest": "assembly/6a2bd71840ef97020185c6357b101a70.zip",
          "src": "staging/6a2bd71840ef97020185c6357b101a70",
        },
        "type": "ZipDirectoryTask",
      },
    },
  },
  "staging": Object {
    "6a2bd71840ef97020185c6357b101a70": Object {
      "external-link.txt": "this is a normal file",
      "file1.txt": "file1",
      "local-link.txt": "file1",
      "subdir": Object {
        "file2.txt": "file2 in subdir",
      },
      "subdir2": Object {
        "empty-subdir": Object {
          ".hidden": "hidden file",
        },
        "subdir3": Object {
          "file3.txt": "file3 in subdir2/subdir3",
        },
      },
    },
  },
}
`;

exports[`zip directory assets 1`] = `
Object {
  "assembly": Object {
    "manifest.json": Object {
      "artifacts": Object {
        "asset_6a2bd71840ef97020185c6357b101a70": Object {
          "environment": "aws://unknown-account/unknown-region",
          "properties": Object {
            "file": "6a2bd71840ef97020185c6357b101a70.zip",
          },
          "type": "aws:s3:object",
        },
        "stack1": Object {
          "environment": "aws://unknown-account/unknown-region",
          "properties": Object {
            "parameters": Object {
              "asset6a2bd71840ef97020185c6357b101a70S3Bucket": "\${asset_6a2bd71840ef97020185c6357b101a70}.s3bucket",
              "asset6a2bd71840ef97020185c6357b101a70S3ContentHash": "\${asset_6a2bd71840ef97020185c6357b101a70}.sha256",
              "asset6a2bd71840ef97020185c6357b101a70S3Key": "\${asset_6a2bd71840ef97020185c6357b101a70}.s3key",
            },
            "templateFile": "stack1.template.json",
          },
          "type": "aws:cloudformation:stack",
        },
      },
      "runtime": Object {
        "libraries": Object {
          "jsii-runtime": "node.js/v10.13.0",
        },
      },
      "version": "0.19.0",
    },
    "stack1.template.json": Object {
      "Parameters": Object {
        "asset6a2bd71840ef97020185c6357b101a70S3Bucket": Object {
          "Description": "S3 bucket for asset 6a2bd71840ef97020185c6357b101a70 from /Users/benisrae/code/cdk/aws-cdk-3/packages/@aws-cdk/assets/test/fixtures/test1",
          "Type": "String",
        },
        "asset6a2bd71840ef97020185c6357b101a70S3ContentHash": Object {
          "Description": "SHA-256 hash of the asset as it was uploaded to S3",
          "Type": "String",
        },
        "asset6a2bd71840ef97020185c6357b101a70S3Key": Object {
          "Description": "S3 key for asset 6a2bd71840ef97020185c6357b101a70 from /Users/benisrae/code/cdk/aws-cdk-3/packages/@aws-cdk/assets/test/fixtures/test1",
          "Type": "String",
        },
      },
    },
  },
  "build.json": Object {
    "steps": Object {
      "asset_6a2bd71840ef97020185c6357b101a70": Object {
        "parameters": Object {
          "dest": "assembly/6a2bd71840ef97020185c6357b101a70.zip",
          "src": "staging/6a2bd71840ef97020185c6357b101a70",
        },
        "type": "ZipDirectoryTask",
      },
    },
  },
  "staging": Object {
    "6a2bd71840ef97020185c6357b101a70": Object {
      "external-link.txt": "this is a normal file",
      "file1.txt": "file1",
      "local-link.txt": "file1",
      "subdir": Object {
        "file2.txt": "file2 in subdir",
      },
      "subdir2": Object {
        "empty-subdir": Object {
          ".hidden": "hidden file",
        },
        "subdir3": Object {
          "file3.txt": "file3 in subdir2/subdir3",
        },
      },
    },
  },
}
`;
